cirq/google/op_serializer.py

```diff
@@ -0,0 +1,174 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, cast, Dict, List, NamedTuple, Type, TypeVar, Union
+
+import sympy
+
+from cirq import devices, ops
+from cirq.google import arg_func_langs
+
+# Type for variables that are subclasses of ops.Gate.
+Gate = TypeVar('Gate', bound=ops.Gate)
+
+
+class SerializingArg(
+        NamedTuple('SerializingArg',
+                   [('serialized_name', str),
+                    ('serialized_type', Type[arg_func_langs.ArgValue]),
+                    ('gate_getter',
+                     Union[str, Callable[[ops.Gate], arg_func_langs.ArgValue]]),
+                    ('required', bool)])):
+    """Specification of the arguments for a Gate and its serialization.
+
+    Attributes:
+        serialized_name: The name of the argument when it is serialized.
+        serialized_type: The type of the argument when it is serialized.
+        gate_getter: The name of the property or attribute for getting the
+            value of this argument from a gate, or a function that takes a
+            gate and returns this value. The later can be used to supply
+            a value of the serialized arg by supplying a lambda that
+            returns this value (i.e. `lambda x: default_value`)
+        required: Whether this argument is a required argument for the
+            serialized form.
+    """
+
+    def __new__(cls,
+                serialized_name,
+                serialized_type,
+                gate_getter,
+                required=True):
+        return super(SerializingArg,
+                     cls).__new__(cls, serialized_name, serialized_type,
+                                  gate_getter, required)
+
+
+class GateOpSerializer:
+    """Describes how to serialize a GateOperation for a given Gate type.
+
+    Attributes:
+        gate_type: The type of the gate that can be serialized.
+        serialized_gate_id: The id used when serializing the gate.
+    """
+
+    def __init__(self, *, gate_type: Type[Gate], serialized_gate_id: str,
+                 args: List[SerializingArg]):
+        """Construct the serializer.
+
+        Args:
+            gate_type: The type of the gate that is being serialized.
+            serialized_gate_id: The string id of the gate when serialized.
+            args: A list of specification of the arguments to the gate when
+                serializing, including how to get this information from the
+                gate of the given gate type.
+        """
+        self.gate_type = gate_type
+        self.serialized_gate_id = serialized_gate_id
+        self.args = args
+
+    def to_proto(self, op: ops.GateOperation) -> Dict:
```

Comment:
Could this instead produce actual proto message instances? There could be a parallel `to_proto_dict` method (or `to_json`) that produces a json-compatible dict (I presume the conversion could be automatic since there is a canonical json encoding for proto v3).

Reply:
We could but I'm not sure of the utility.  The python proto library is kind of a mess and we will just end up passing this as a protodict to engine code (I don't think it is ever in proto message form in Cirq right now).

Reply:
One advantage of proto classes is that we could use something like ['mypy-protobuf'](https://github.com/dropbox/mypy-protobuf) to generate `.pyi` stubs and then mypy can typecheck the code.

Reply:
That would be nice.  I'm going to leave as proto dicts for now.  I've relabeled the methods to say proto_dict in the serializer and deserializer.